builder/comp-builder.nix

```diff
@@ -166,14 +163,59 @@ let
   exeName = componentId.cname + exeExt;
   testExecutable = "dist/build/${componentId.cname}/${exeName}";
 
-in stdenv.lib.fix (drv:
+  # Attributes that are common to both the build and haddock derivations
+  commonAttrs = {
+      src = cleanSrc;
+
+      LANG = "en_US.UTF-8";         # GHC needs the locale configured during the Haddock phase.
+      LC_ALL = "en_US.UTF-8";
+
+      enableParallelBuilding = true;
+
+      SETUP_HS = setup + /bin/Setup;
+
+      prePatch = if (cabalFile != null)
+         then ''cat ${cabalFile} > ${package.identifier.name}.cabal''
+         else lib.optionalString (cabal-generator == "hpack") ''
+           ${buildPackages.haskell-nix.haskellPackages.hpack.components.exes.hpack}/bin/hpack
+         '';
+    }
+    # patches can (if they like) depend on the version and revision of the package.
+    // lib.optionalAttrs (patches != []) {
+      patches = map (p:
+        if builtins.isFunction p
+          then p { inherit (package.identifier) version; inherit revision; }
+          else p
+        ) patches;
+    }
+    // haskellLib.optionalHooks {
+      inherit preUnpack postUnpack preConfigure postConfigure
+        preBuild postBuild preHaddock postHaddock
+        preInstall postInstall;
```

Comment:
*Should* we get the build/install hooks in the haddock builder? I'd expect most uses of those to only work for the main component (e.g. generating `optparse-applicative` completions in `postInstall`).

Reply:
My inclination would be to share the unpack and configure hooks, haddock hook only for haddock, and the others only for the main build.

Reply:
I have split the build specific and haddock specific hooks out of `commonAttrs` and added them to just the derivation they are needed for.